pkg/fleets/controller.go

```diff
@@ -309,19 +314,119 @@ func (c *Controller) deleteEmptyGameServerSets(fleet *stablev1alpha1.Fleet, list
 }
 
 // recreateDeployment applies the recreate deployment strategy to all non-active
-// GameServerSets
-func (c *Controller) recreateDeployment(list []*stablev1alpha1.GameServerSet) error {
-	for _, gsSet := range list {
+// GameServerSets, and return the replica count for the active GameServerSet
+func (c *Controller) recreateDeployment(fleet *stablev1alpha1.Fleet, rest []*stablev1alpha1.GameServerSet) (int32, error) {
+	for _, gsSet := range rest {
 		if gsSet.Spec.Replicas != 0 {
 			c.logger.WithField("gameserverset", gsSet.ObjectMeta.Name).Info("applying recreate deployment: scaling to 0")
 			gsSetCopy := gsSet.DeepCopy()
 			gsSetCopy.Spec.Replicas = 0
 			if _, err := c.gameServerSetGetter.GameServerSets(gsSetCopy.ObjectMeta.Namespace).Update(gsSetCopy); err != nil {
-				return errors.Wrapf(err, "error updating gameserverset %s", gsSetCopy.ObjectMeta.Name)
+				return 0, errors.Wrapf(err, "error updating gameserverset %s", gsSetCopy.ObjectMeta.Name)
 			}
+			c.recorder.Eventf(fleet, corev1.EventTypeNormal, "ScalingGameServerSet",
+				"Scaling inactive GameServerSet %s from %d to %d", gsSetCopy.ObjectMeta.Name, gsSet.Spec.Replicas, gsSetCopy.Spec.Replicas)
 		}
 	}
 
+	return fleet.LowerBoundReplicas(fleet.Spec.Replicas - stablev1alpha1.SumStatusAllocatedReplicas(rest)), nil
+}
+
+// rollingUpdateDeployment will do the rolling update of the old GameServers
+// through to the new ones, based on the fleet.Spec.Strategy.RollingUpdate configuration
+// and return the replica count for the active GameServerSet
+func (c *Controller) rollingUpdateDeployment(fleet *stablev1alpha1.Fleet, active *stablev1alpha1.GameServerSet, rest []*stablev1alpha1.GameServerSet) (int32, error) {
+	replicas, err := c.rollingUpdateActive(fleet, active, rest)
+	if err != nil {
+		return replicas, err
+	}
+	if err := c.rollingUpdateRest(fleet, rest); err != nil {
+		return replicas, err
+	}
+	return replicas, nil
+}
+
+// rollingUpdateActive applies the rolling update to the active GameServerSet
+// and returns what its replica value should be set to
+func (c *Controller) rollingUpdateActive(fleet *stablev1alpha1.Fleet, active *stablev1alpha1.GameServerSet, rest []*stablev1alpha1.GameServerSet) (int32, error) {
+	replicas := active.Spec.Replicas
+
+	// if the active spec replicas are greater than or equal the fleet spec replicas, then we don't
```

Comment:
@enocom made the inversion here, and also put in some text to help explain the logic. WDYT?